File: Algorithm/Greedy/Huffman Coding/algo.js
class Node {
  constructor(freq, char = null, left = null, right = null) {
    this.freq = freq;
    this.char = char;
    this.left = left;
    this.right = right;
  }
}

function buildHuffmanTree(text) {
  const freqMap = new Map();
  for (const char of text) {
    freqMap.set(char, (freqMap.get(char) || 0) + 1);
  }

  const heap = Array.from(freqMap.entries()).map(([char, freq]) => new Node(freq, char));
  heap.sort((a, b) => a.freq - b.freq);

  while (heap.length > 1) {
    const left = heap.shift();
    const right = heap.shift();
    const newNode = new Node(left.freq + right.freq, null, left, right);
    heap.unshift(newNode);
    heap.sort((a, b) => a.freq - b.freq);
  }

  return heap[0];
}

function buildCodes(node, prefix = '', codeMap = {}) {
  if (node.char) {
    codeMap[node.char] = prefix;
  } else {
    buildCodes(node.left, prefix + '0', codeMap);
    buildCodes(node.right, prefix + '1', codeMap);
  }
  return codeMap;
}

function huffmanEncoding(text) {
  const root = buildHuffmanTree(text);
  const huffmanCode = buildCodes(root);
  return huffmanCode;
}

const text = "huffman coding is a data compression algorithm";
const huffmanCode = huffmanEncoding(text);
for (const [char, code] of Object.entries(huffmanCode)) {
  console.log(`${char}: ${code}`);
}

File: Algorithm/Greedy/Huffman Coding/algo.py
import heapq
from collections import defaultdict, Counter

class Node:
    def __init__(self, freq, char=None, left=None, right=None):
        self.freq = freq
        self.char = char
        self.left = left
        self.right = right

    def __lt__(self, other):
        return self.freq < other.freq

def build_huffman_tree(text):
    frequency = Counter(text)
    heap = [Node(freq, char) for char, freq in frequency.items()]
    heapq.heapify(heap)

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)
        merged = Node(left.freq + right.freq, left=left, right=right)
        heapq.heappush(heap, merged)

    return heap[0]

def build_codes(node, prefix='', code_map={}):
    if node.char:
        code_map[node.char] = prefix
    else:
        build_codes(node.left, prefix + '0', code_map)
        build_codes(node.right, prefix + '1', code_map)
    return code_map

def huffman_encoding(text):
    root = build_huffman_tree(text)
    huffman_code = build_codes(root)
    return huffman_code

if __name__ == "__main__":
    text = "huffman coding is a data compression algorithm"
    huffman_code = huffman_encoding(text)
    for char, code in huffman_code.items():
        print(f"{char}: {code}")
